src/main/java/org/jenkinsci/plugins/mesos/MesosApi.java

```diff
@@ -49,18 +65,24 @@ public MesosApi(String masterUrl, String user, String frameworkName)
             .withValue("master-url", ConfigValueFactory.fromAnyRef(masterUrl));
     this.clientSettings = MesosClientSettings.fromConfig(conf);
     system = ActorSystem.create("mesos-scheduler");
+    context = system.dispatcher();
     materializer = ActorMaterializer.create(system);
-
     client = connectClient().get();
+    stateMap = new ConcurrentHashMap<>();
+    adapter = new SchedulerAdapter(Scheduler.fromClient(client), materializer, context);
+    adapterSourceSink = adapter.asSourceAndSink(SpecsSnapshot.empty());
+    adapterSourceSink.t2().runForeach(this::updateState, materializer);
+    adapterQueues = adapter.asAkkaQueues(SpecsSnapshot.empty(), OverflowStrategy.backpressure());
   }
 
   /**
-   * Start a Jekins slave/agent on Mesos.
-   *
-   * @return A future reference to the {@link MesosSlave}.
+   * Enqueue spec for a jenkins agent that will eventually come online.
    */
-  public CompletableFuture<MesosSlave> startAgent() {
-    throw new NotImplementedException();
+  public void enqueueAgent() {
+    PodSpec spec = buildMesosAgentTask(0.1, 32);
```

Comment:
@jeschkies dont think we will be able to return a CompleteableFuture<MesosSlave> here since after enqueue all we have is a CompletionStage<QueueOfferResult> and we have not created a MesosSlave at this point,

this could still work based on https://github.com/jenkinsci/nomad-plugin/blob/master/src/main/java/org/jenkinsci/plugins/nomad/NomadApi.java#L33

Reply:
You are right. However, in `MesosCloud` we will have to "wait" for the task to be running and the agent to be connect. We can handle this with `CompletableFuture`.